server/src/main/java/org/apache/druid/server/coordination/ServerManager.java

```diff
@@ -258,21 +259,23 @@ protected <T> QueryRunner<T> buildQueryRunnerForSegment(
       return new ReportTimelineMissingSegmentQueryRunner<>(descriptor);
     }
 
-    final ReferenceCountingSegment segment = chunk.getObject();
-    return buildAndDecorateQueryRunner(
+    final ReferenceCountedSegmentProvider referenceCounter = chunk.getObject();
+
+    final Optional<Segment> maybeSegment = segmentMapFn.apply(referenceCounter);
```

Comment:
I think with this change, we're now acquiring references prior to creation of the `QueryRunner` for a segment; where previously they were acquired on the call to `QueryRunner#run`. It means we could leak references in two new cases:

- if an exception occurs in `buildAndDecorateQueryRunner`
- if the query runner is never actually run

The first thing can be fixed by wrapping the call to `buildAndDecorateQueryRunner` in a `catch (Throwable e)` and closing the `seg` if it fails.

The second is trickier. I think in some cases it does make sense for query runners to not ever be run. For example, an unordered `scan` query with `limit` could exit early before processing all segments.

Reply:
good catch, I didn't actually mean to change this flow as part of this PR - i was doing some experiments in a branch on top of this involving acquiring the references up front, but this was a strange intermediary state and not even what that branch looks like right now (it pushes in a closer that the segments can register with to make it easier to handle the issues you're mentioning), I think i got branches mixed up. 

I have switched this to what I meant to do for now, which is push the provider and map function into `ReferenceCountingSegmentQueryRunner`, which keeps the reference handling and closing inside the runner